api/views/workflow.py

```diff
@@ -40,10 +44,86 @@ class SourceViewSet(viewsets.ReadOnlyModelViewSet):
     ordering = ('id', )
     ordering_fields = ('long_title', )
     serializer_class = SourceSerializer
+    permission_classes = [DjangoModelPermissionsOrAnonReadOnly, ]
+
+    queryset = Source.objects.none()  # Required for DjangoModelPermissions
+
+    VALID_IMAGE_TYPES = ('image/png', 'image/jpeg')
 
     def get_queryset(self):
         return Source.objects.exclude(icon='').exclude(is_deleted=True)
 
+    def create(self, request, *args, **kwargs):
+        try:
+            long_title = request.data['long_title']
+            icon = request.data['icon']
+        except KeyError as e:
+            raise ValidationError('{} is a required attribute.'.format(e))
+
+        try:
+            r = requests.get(icon, timeout=5)
+            header_type = r.headers['content-type'].split(';')[0].lower()
+            if header_type not in self.VALID_IMAGE_TYPES:
+                raise ValidationError('Invalid type. Expected one of {}. Received {}'.format(self.VALID_IMAGE_TYPES, header_type))
+
+            icon_file = ContentFile(r.content)
+        except Exception as e:
+            raise ValidationError('Could not download/process image. {}'.format(e))
```

Comment:
Can you throw a `logger.exception` in here?

I don't think we should send back the exception here. It could technically be possible to leak data if we receive a chatty exception.